onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -231,37 +231,29 @@ static Status ForceSingleNodeCPUFloat16ToFloat32(onnxruntime::Graph& graph, cons
   return Status::OK();
 }
 
-enum TypeGroup {
-  Unknown = -1,
-  Bool = 0,
-  Integer = 1,
-  Float = 2,
-};
-
-TypeGroup GetTypeGroup(DataType type) {
-  if (*type == "tensor(bool)") {
-    return Bool;
-  }
-
-  if (*type == "tensor(int16)" || *type == "tensor(int32)" || *type == "tensor(int64)" || *type == "tensor(int8)" ||
-      *type == "tensor(uint16)" || *type == "tensor(uint32)" || *type == "tensor(uint64)" || *type == "tensor(uint8)") {
-    return Integer;
-  }
-
-  if (*type == "tensor(bfloat16)" || *type == "tensor(double)" || *type == "tensor(float)" || *type == "tensor(float16)") {
-    return Float;
-  }
-
-  return Unknown;
-}
-
 /** Transformer to remove duplicate Cast nodes. */
 class RemoveDuplicateCastTransformer : public GraphTransformer {
  public:
   RemoveDuplicateCastTransformer() : GraphTransformer("RemoveDuplicateCastTransformer") {
   }
 
  private:
+  static constexpr std::array<std::string_view, 13> cast_ordering{
+      "tensor(bool)", "tensor(uint8)", "tensor(uint16)", "tensor(uint32)", "tensor(uint64)", "tensor(int8)", "tensor(int16)",
+      "tensor(int32)", "tensor(int64)", "tensor(bfloat16)", "tensor(float16)", "tensor(float)", "tensor(double)"};
```

Comment:
There will be precision loss in both direction: bfloat16 to float16, and float16 to bfloat16.

Reply:
Actually, it is more complicated since we might also need consider the original data type. For example,
(original type = bool)  ---> Cast(to=int8) ---> Cast(to=float): it is safe to remove the Cast(to=int8)
(original type = int8)  ---> Cast(to=int32) ---> Cast(to=float): it is safe to remove the Cast(to=int32)
(original type = float) ---> Cast(to=int32) ---> Cast(to=float): it is not safe to remove the Cast(to=int32)

Reply:
Is this not just along the lines of the comment the previous author put in the code:

https://github.com/microsoft/onnxruntime/blob/35ecce45496ee752bc3f85618eba713bb50e6069/onnxruntime/core/optimizer/insert_cast_transformer.cc#L281-L288

This function is just determining a singular cast node loses precision so if either subsequent cast loses precision we should not remove the cast operator.